test/lib/auth_cache/assume_role_credentials_adapter_test.exs

```diff
@@ -0,0 +1,55 @@
+defmodule ExAws.STS.AuthCache.AssumeRoleCredentialsAdapterTest do
+  use ExUnit.Case, async: true
+  alias ExAws.STS.AuthCache.AssumeRoleCredentialsAdapter
+
+  import Mox
+
+  test "#adapt_auth_config" do
```

Comment:
I usually have the `test` block describe what are the expectation for the current test. Also, following the mantra of a single assert per test, I'd break this test into two, one to verify we call the HTTP module with the right arguments, and another to verify the return value. Something like:
```elixir
test "adapt_auth_config/2 calls the HTTP client with the right arguments" do
  body = %{
    access_key_id: "1",
    secret_access_key: "secret",
    session_token: "token"
  }

  ExAws.Request.HttpMock
  |> expect(:request, fn _method,
                           _url,
                           "Action=AssumeRole&DurationSeconds=900&RoleArn=1111111%2Ftest_role&RoleSessionName=test&Version=2011-06-15",
                           _headers,
                           _opts ->
      {:ok, %{status_code: 200, body: body}}
    end)

  AssumeRoleCredentialsAdapter.adapt_auth_config(profile, 300, &test_loader/1)
end

test "adapt_auth_config/2 returns a map with the expected fields" do
  body = %{
    access_key_id: "1",
    secret_access_key: "secret",
    session_token: "token"
  }

  ExAws.Request.HttpMock
  |> stub(:request, fn _method,
                           _url,
                           "Action=AssumeRole&DurationSeconds=900&RoleArn=1111111%2Ftest_role&RoleSessionName=test&Version=2011-06-15",
                           _headers,
                           _opts ->
      {:ok, %{status_code: 200, body: body}}
    end)

  auth = test_loader("default")


  expected = %{
    access_key_id: body.access_key_id,
    secret_access_key: body.secret_access_key,
    security_token: body.session_token,
    role_arn: auth.role_arn,
    role_session_name: auth.role_session_name,
    source_profile: auth.source_profile
  }

  assert expected ==
           AssumeRoleCredentialsAdapter.adapt_auth_config(profile, 300, &test_loader/1)
end
```
:warning: Didn't test this code, beware of copy-pasta :spaghetti: errors :sweat_smile:

Reply:
Not sure if I fully agree in this case, as the goal with the http mock is more to make sure it is specific enough I won't be mistakenly giving `200 OK` answers to other requests the library may fire (and with the fallback to use :instance_role on this lib, this could hide errors in the subject under test).

The sole assertion is really the true expectation for this test.

Reply:
Yeah, in this case since you're getting something back (and your code depends on it), you're implicitly validating everything with a single expectation. I see HTTP requests (or any outside request) as side effects, and then usually follow Sandi Metz's advice:

![](https://miro.medium.com/proxy/1*_xBySVnBsfLSDHGl1DXkeA.png)

In this case this is more of outgoing query than a command, so disregard my previous comment :+1:

Reply:
Yeah, that matrix is a great reference and it was also what informed this particular decision 👍